### PR TITLE
Fennec history import

### LINKED ISSUE #1460: Add Fennec history visits import code
Opening this early as I'm working on a prototype.

Here's roughly the plan:
- Android Components will arrange to call us with a path to the legacy `browser.db` file. We can attach that database and run SQL queries on it to import the data in our database.
- [Fennec history/visits schema](https://gist.github.com/eoger/6473274943d9219df1ff2815db0538a8)
- [Fenix->Fennec mapping](https://gist.github.com/eoger/88854d29cff178c545266d4d9c7f72fc) (thanks @linacambridge for reviewing it!)

┆Issue is synchronized with this [Jira Story](https://jira.allizom.org/browse/SYNC-211)
┆Story Points: 5
┆Sprint: SYNC - 2019-08-30
┆Epic: [Q3 2.1] Sync Migration

### PR DESCRIPTION
**Pull request by eoger**
Fixes #1460.
It's still missing some more complex tests though, but I'd be nice to get some eyes on this early.

### CODE REVIEW

File: components/places/src/import/common.rs [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+            Ok(None)
+        }
+    }
+
+    #[inline(never)]
+    pub fn is_valid_url(ctx: &Context<'_>) -> Result<Option<bool>> {
+        Ok(match ctx.get_raw(0) {
+            ValueRef::Text(s) if s.len() <= URL_LENGTH_MAX => Some(Url::parse(s).is_ok()),
+            // Should we do this?
+            // ValueRef::Null => None,
```
Comment (linacambridge) [inline components/places/src/import/common.rs]: Eh...I don't think so. It'd be consistent with how SQL functions handle `NULL`s, but I don't think we gain much from distinguishing between missing and invalid.

File: components/places/src/import/fennec/history.rs [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+            FROM fennec.history h
+            WHERE is_valid_url(h.url)"
+    ;
+
+    // Insert history visits
+    static ref INSERT_HISTORY_VISITS: &'static str =
+        "INSERT OR IGNORE INTO main.moz_historyvisits(from_visit, place_id, visit_date, visit_type, is_local)
+            SELECT
+                NULL,
+                (SELECT p.id FROM main.moz_places p WHERE p.url_hash = hash(h.url)),
```
Comment (linacambridge) [inline components/places/src/import/fennec/history.rs]: The hash function isn't unique, so I think we need a `p.url = h.url` check, too.

File: components/places/tests/fennec_history.rs [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+    let v = format!("test{}_______", c);
+    let s = &v[..12];
+    Guid::from(s)
+}
+
+impl Default for FennecHistory {
+    fn default() -> Self {
+        Self {
+            title: None,
+            url: "".to_owned(),
```
Comment (linacambridge) [inline components/places/tests/fennec_history.rs]: Nit: `String::new()`

File: components/places/src/import/fennec/history.rs [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+                sanitize_timestamp(v.date),
+                v.visit_type,
```
Comment (linacambridge) [inline components/places/src/import/fennec/history.rs]: Same here, a comment that Fennec's visit types map 1:1 to ours would be great!

File: components/places/src/import/fennec/history.rs [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+                1
+            FROM fennec.history h
+            WHERE is_valid_url(h.url)"
+    ;
+
+    // Insert history visits
+    static ref INSERT_HISTORY_VISITS: &'static str =
+        "INSERT OR IGNORE INTO main.moz_historyvisits(from_visit, place_id, visit_date, visit_type, is_local)
+            SELECT
+                NULL,
```
Comment (linacambridge) [inline components/places/src/import/fennec/history.rs]: A comment that Fennec doesn't store enough info to rebuild redirect chains (we know if something was a redirect target from its transition type, but we don't know the source) would be wonderful, to remind ourselves why this is `NULL` later.

**Comment by eoger** [comment]
Thanks @linacambridge I've addressed your review comments and added more tests.

File: components/places/android/src/main/java/mozilla/appservices/places/PlacesConnection.kt [hunk 6]
```diff
@@ -740,6 +747,16 @@ interface WritableHistoryConnection : ReadableHistoryConnection {
...
 
+    /**
+     * Imports visits from a Fennec `browser.db` database.
+     *
+     * It has been designed exclusively for non-sync users and should
+     * be called before bookmarks import.
+     *
+     * @param path Path to the `browser.db` file database.
+     */
+    fun importVisitsFromFennec(path: String)
```
Comment (vladikoff) [inline components/places/android/src/main/java/mozilla/appservices/places/PlacesConnection.kt]: 🎆

**Pull request merged**